packages/@lwc/engine-core/src/framework/base-lightning-element.ts

```diff
@@ -224,19 +219,49 @@ export const LightningElement: LightningElementConstructor = function (
 
     // Linking elm, shadow root and component with the VM.
     associateVM(component, vm);
-    associateVM(cmpRoot, vm);
     associateVM(elm, vm);
 
     // Adding extra guard rails in DEV mode.
     if (process.env.NODE_ENV !== 'production') {
         patchCustomElementWithRestrictions(elm);
         patchComponentWithRestrictions(component);
-        patchShadowRootWithRestrictions(cmpRoot);
+    }
+
+    if (!features.ENABLE_LIGHT_DOM_COMPONENTS) {
+        assert.isTrue(
+            ctor.shadow,
+            `${ctor.name} is a Light DOM component but Light DOM components are disabled. Either make the component a regular component (by setting \`static shadow = false\`) or set the runtime flag \`ENABLE_LIGHT_DOM_COMPONENTS\`.`
+        );
+    }
+
+    if (ctor.shadow) {
```

Comment:
The `ctor.shadow` property is a user-defined value, we should certainly be more careful with this value. The current check will coerce the `shadow` value which means that `shadow = ''` and `shadow = null` will both make the component render in the light DOM.

I would recommend asserting that `ctor.shadow` is a boolean before doing anything with it.